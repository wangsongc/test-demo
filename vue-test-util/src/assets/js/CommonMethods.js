/**
 * 获取当前视口的宽度，判断默认样式
 * @returns true<1025<=false
 */
export function checkWidth (size) {
  let width = size || window.document.body.offsetWidth
  if (width < 1025) {
    return true
  } else {
    return false
  }
}
