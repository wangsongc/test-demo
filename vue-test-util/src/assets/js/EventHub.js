// (C) 2018 SEIKO SOLUTIONS INC.

import Vue from 'vue'

const eventHub = new Vue()

export default eventHub
