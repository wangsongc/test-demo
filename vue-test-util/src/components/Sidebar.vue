<template>
  <div class="sidebar">
    <el-menu
      class="sidebar-el-menu"
      :default-active="onRoutes"
      :collapse="collapse"
      background-color="#324157"
      text-color="white"
      active-text-color="#20a0ff"
      unique-opened
      router
      collapse-transition
    >
      <template v-for="item in items">
        <template v-if="item.subs">
          <el-submenu :index="item.index" :key="item.index">
            <template slot="title">
              <i :class="item.icon"></i>
              <span slot="title" style="margin-left: 10px">{{ item.title }}</span>
            </template>
            <el-menu-item
              v-for="(subItem, i) in item.subs"
              :key="i"
              :index="subItem.index"
              style="margin-left: 10px"
            >
              <i :class="subItem.icon"></i>
              <span slot="title" style="margin-left: 10px">{{ subItem.title }}</span>
            </el-menu-item>
          </el-submenu>
        </template>
        <template v-else>
          <el-menu-item :index="item.index" :key="item.index">
            <i :class="item.icon"></i>
            <span slot="title" style="margin-left: 10px" color="white">{{ item.title }}</span>
          </el-menu-item>
        </template>
      </template>
    </el-menu>
  </div>
</template>

<script>
import eventHub from '@/assets/js/EventHub.js'
import { checkWidth } from '@/assets/js/CommonMethods.js'

export default {
  data () {
    return {
      collapse: checkWidth(),
      items: [
        {
          icon: 'el-icon-tickets',
          index: 'overview',
          title: '系统概览'
        },
        {
          icon: 'el-icon-sort',
          index: 'user',
          title: '用户管理',
          subs: [
            {
              icon: 'el-icon-arrow-right',
              index: 'users',
              title: '普通用户'
            },
            {
              icon: 'el-icon-d-arrow-right',
              index: 'vips',
              title: 'VIP 用户'
            }
          ]
        }
      ]
    }
  },
  computed: {
    onRoutes () {
      return this.$route.path.replace('/', '')
    }
  },
  created () {
    // 通过 Event Bus 进行组件间通信，来折叠侧边栏
    eventHub.$on('collapse', msg => {
      this.collapse = msg
    })
  }
}
</script>

<style scoped>
.sidebar {
  display: block;
  position: absolute;
  left: 0;
  top: 70px;
  bottom: 0;
}

.sidebar-el-menu:not(.el-menu--collapse) {
  width: 200px;
}

.sidebar > ul {
  height: 100%;
}
</style>
