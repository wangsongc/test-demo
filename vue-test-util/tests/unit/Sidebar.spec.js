/**
 * SideBar.vue组件
 * 测试内容包括以下
 * func
 * element-ui插件的使用
 * $route（伪造$route便于测试计算属性）
 * computed
 * DOM
 */
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Element from 'element-ui'
import Sidebar from '@/components/Sidebar.vue'
import eventHub from '@/assets/js/EventHub.js'
import { checkWidth } from '@/assets/js/CommonMethods.js'

// 创建临时Vue实例，挂载组件中使用的插件
const localVue = createLocalVue()
localVue.use(Element)
// 伪造$route,用于测试计算属性
const $route = {
  path: '/vips'
}

describe('Sidebar.vue', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallowMount(Sidebar, { localVue, mocks: { $route } })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  // 测试内容：data中的函数checkWidth()
  it('checkWidth test', () => {
    // window.document.body.offsetWidth的值再测试环境下一直都是0，所以checkWidth的值一致都是true，无法改变
    expect(checkWidth()).toBeTruthy()
    expect(wrapper.vm.collapse).toBeTruthy()
  })

  // 测试内容：computed计算属性onRoutes
  it('computed', () => {
    expect(wrapper.vm.onRoutes).toBe('vips')
  })

  // 测试内容：eventHub $on
  it('eventHub $on test', () => {
    wrapper.vm.collapse = false
    eventHub.$emit('collapse', true)
    expect(wrapper.vm.collapse).toBeTruthy()
  })

  // 测试内容：snapshot->概括的测试DOM结构
  // 如果组件内存在比较特殊的需要测试的DOM结构的话，可以单独测试（详见AppButton测试文件末尾）
  it('snapshot test', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
