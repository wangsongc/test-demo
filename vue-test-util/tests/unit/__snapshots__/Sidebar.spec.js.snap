// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar.vue snapshot test 1`] = `
<div class="sidebar">
  <el-menu-stub mode="vertical" defaultactive="vips" uniqueopened="true" router="true" menutrigger="hover" collapse="true" backgroundcolor="#324157" textcolor="white" activetextcolor="#20a0ff" collapsetransition="true" class="sidebar-el-menu">
    <el-menu-item-stub index="overview"><i class="el-icon-tickets"></i> <span color="white" style="margin-left: 10px;">系统概览</span></el-menu-item-stub>
    <el-submenu-stub offset="0" boundariespadding="5" popperoptions="[object Object]" index="user" showtimeout="300" hidetimeout="300"><template></template>
      <el-menu-item-stub index="users" style="margin-left: 10px;"><i class="el-icon-arrow-right"></i> <span style="margin-left: 10px;">普通用户</span></el-menu-item-stub>
      <el-menu-item-stub index="vips" style="margin-left: 10px;"><i class="el-icon-d-arrow-right"></i> <span style="margin-left: 10px;">VIP 用户</span></el-menu-item-stub>
    </el-submenu-stub>
  </el-menu-stub>
</div>
`;
